Replicator/ReplicatorOptions.hh

```diff
@@ -303,24 +303,37 @@ namespace litecore { namespace repl {
             for (auto& opt : collectionOpts) {
                 specs.emplace_back(collectionPathToSpec(opt.collectionPath));
             }
-            for (size_t i = 0; i < activeCollections.size(); ++i) {
-                auto findIt = _collectionSpecToIndex.find(activeCollections[i]);
-                if (findIt == _collectionSpecToIndex.end()) {
+
+            for (size_t activeIndex = 0; activeIndex < activeCollections.size(); ++activeIndex) {
+                auto foundEntry = _collectionSpecToIndex.find(activeCollections[activeIndex]);
+                if (foundEntry == _collectionSpecToIndex.end()) {
                     // Put the unfound spec to the end of collectionOpts
-                    size_t j = collectionOpts.size();
+                    size_t passiveCount = collectionOpts.size();
                     collectionOpts.emplace_back(nullslice);
-                    // Assertion: collectionOpts[j].collectionPath == nullslice
-                    _collectionSpecToIndex[specs[i]] = j;
-                    std::swap(collectionOpts[i], collectionOpts[j]);
+                    if (activeIndex < origCount) {
+                        // Assertion: collectionOpts[j].collectionPath == nullslice
+                        // We haven't run out of passive entries yet, so make sure
+                        // the nullslice one is in the correct place alongside
+                        // its active counterpart.  The spec to index dictionary will
+                        // point to the out of bounds, and unusable, original entry
+                        // for the spec.
+                        _collectionSpecToIndex[specs[activeIndex]] = passiveCount;
+                        std::swap(collectionOpts[activeIndex], collectionOpts[passiveCount]);
+                    }
```

Comment:
we need the "else",
else {
    auto path = collectionOpts[activeIndex].collectionPath;
    if (path) {
        auto spec = Options::collectionPathToSpec(path);
       _collectionSpecToIndex[spec] = passiveCount;
    }
}
 
Looking at this example,
activeCollections:  A B C
this->collections:  C
We cannot lose track of C